### ISSUE TITLE
Intersect ray hit RID is different each time?

### ISSUE THREAD

**Issue by gitcatrat** [godotengine/godot#36824]
**Godot version:** 
3.2.stable.mono

**OS/device including version:**
macOS Catalina 10.15.3/MacBook Pro

**Issue description:**
`DirectSpaceState.IntersectRay` hit object `RID` is different each time, even if same hit body.

I didn't even notice this before because I used nodes and `collider` (and its `Name`) to identify the hit target but now that I've moved some of my stuff to use Servers API directly, I have no way to identify the hit target if `RID` is constantly changing.

Maybe it's supposed to change? I have no idea, couldn't find **any** information about it. I was going to use `RID` as a dictionary key to get my data instance but this condition fails:
``` csharp
public static Dictionary<RID, Enemy> enemyRids;
// ...
if (enemyRids.ContainsKey(hitRID)) { /* ... */ }
```

**Steps to reproduce:**
Sorry for C#, I've used GDScript only few times.

```csharp
public override void _PhysicsProcess(float delta) {
  if (doRayCast) {
    var spaceState = GetWorld().DirectSpaceState;
    var hit = spaceState.IntersectRay(from, to);
    Debug.Log(hit["rid"]); // different each time, even if same hit body
    doRayCast = false;
  }
}
```

**Reproduction project:**
[issue.zip](https://github.com/godotengine/godot/files/4295409/issue.zip)

**Screenshot of RIDs (Newtonsoft JSON serializer) when ray-casting same body:**

<img width="227" alt="wtf" src="https://user-images.githubusercontent.com/58787197/76000598-9b6bc900-5f0c-11ea-85f2-9b3018fa2ae5.png">

**Comment by KoBeWi** [comment]
I tried with GDScript and the RID is constant.

Any reproduction project for this?

**Comment by gitcatrat** [comment]
Was about to make one but then I stumbled on this #35897..

**Comment by gitcatrat** [comment]
Managed to solve it by running `NuGet restore`. Project with said issue: [issue.zip](https://github.com/godotengine/godot/files/4295409/issue.zip)

Blue (left) cube is node-based, red (right) cube is added with Servers. Both have the issue.

**Comment by gitcatrat** [comment]
@KoBeWi Managed to solve previously mentioned issue by running `NuGet restore`. 

Project: [issue.zip](https://github.com/godotengine/godot/files/4295409/issue.zip)
Blue (left) cube is node-based, red (right) cube is added with Servers. Both have the issue.

**Comment by KoBeWi** [comment]
Sooo, when you change the debug line to
`Debug.Log((hit["rid"] as RID).GetId());`
you get consistent IDs. So the RID is indeed not changing, but when you print it directly it might be printing some instance data (each RID might be unique object even if they have the same ID).

No idea what is the "NativeInstance

**Comment by gitcatrat** [comment]
Okay, so the bottom line is that RID comparison is not reliable and it can't be used as a key (in dictionary for example). But at least there's a way to get the ID that can be used in these situations. 

Thanks, I'll make a quick test in my project to see if it solves my issue.

**Comment by KoBeWi** [comment]
> Okay, so the bottom line is that RID comparison is not reliable and it can't be used as a key (in dictionary for example).

This is weird actually, because RID is a built-in type, so they should compare properly. It is possible that it's only C# issue, this probably requires some more testing.

**Comment by gitcatrat** [comment]
@KoBeWi Fixed in `Godot 4.0 beta 17`, I have no idea which version fixed it.
I'm not going to close this myself in case you have some other plans.

**Comment by akien-mga** [comment]
Fixed in 4.0.
